added sapyb and unittests

paskino opened the pull request:
Adds `sapyb` method of `DataContainer` and some unit tests.

I think we should remove all `axpby` references in the code as `sapyb` is able to switch from the c-library to pure numpy implementation, which should fix a few problems that have arisen.

Besides, SIRF has already [developed `sapyb` and deprecated `axpby`](https://github.com/SyneRBI/SIRF/blob/0c22d1dcdf5bb7a5135d50e99ec3eca3e9f4d18f/src/common/SIRF.py#L227-L251).

epapoutsellis commented on Wrappers/Python/cil/framework/framework.py:
```diff
@@ -2481,7 +2481,51 @@ def maximum(self, x2, *args, **kwargs):
+        '''
+        do_numpy = False
+        ret_out = False
+        if self.dtype in [complex, numpy.complex, numpy.complex64]:
```
numpy.sctypes['complex']

paskino replied:
updated

epapoutsellis commented on Wrappers/Python/cil/framework/framework.py:
```diff
@@ -2481,7 +2481,51 @@ def maximum(self, x2, *args, **kwargs):
+        It will try to use the CIL C library and default to numpy operations, in case the C library does
+        not handle the types.
+        '''
+        do_numpy = False
```
I think is better `do_c = True` and for `False` make it agnostic. Moreover, a `numba` implementation will be useful because it will work for `numpy`, `cupy` and `dask`

paskino replied:
Maybe `do_c` is clearer.

paskino replied:
updated using `use_c_lib` and added example.

epapoutsellis commented on Wrappers/Python/cil/framework/framework.py:
```diff
@@ -2481,7 +2481,51 @@ def maximum(self, x2, *args, **kwargs):
+        It will try to use the CIL C library and default to numpy operations, in case the C library does
+        not handle the types.
+        '''
+        do_numpy = False
```
Please add an `Example` section in the `docs`.

epapoutsellis commented on Wrappers/Python/test/test_DataContainer.py:
```diff
@@ -931,6 +931,170 @@ def test_axpby7(self):
+        res = numpy.zeros_like(d1.as_array())
+        numpy.testing.assert_array_equal(res, out.as_array())
+
+    def test_sapyb_scalar_f(self):
```
what is `scalar_f`

epapoutsellis commented on Wrappers/Python/test/test_DataContainer.py:
```diff
@@ -931,6 +931,170 @@ def test_axpby7(self):
         numpy.testing.assert_array_equal(res, out.as_array())
 
+
+    def test_sapyb_datacontainer_f(self):
```
Could you rename the methods:

a) all_DataContainer_floats
b) d1_d2_DataContainer_a_b_float
c) d1_d2_DataContainer_complex_a_b_complex

paskino replied:
I don't think the proposed naming is clearer. Here there are 3 methods for floats names already:

1. `test_sapyb_scalar_f` for `a` and `b` scalar
2. `test_sapyb_datacontainer_f`  for `a` and `b` `DataContainer`s
3. `test_sapyb_datacontainer_scalar_f` for `a` and `b` mixed

Plus the respective for complex `_c`.

epapoutsellis commented on Wrappers/Python/test/test_DataContainer.py:
```diff
@@ -931,6 +931,170 @@ def test_axpby7(self):
+        d2.sapyb(b,d1,a, out)
+        numpy.testing.assert_array_equal(res.as_array(), out.as_array())
+
+    def test_sapyb_scalar_c(self):
```
What is the difference with  `test_sapyb_datacontainer_scalar_f`

paskino replied:
`test_sapyb_scalar_c` is for complex only scalar values of the parameters `a` and `b`. `test_sapyb_datacontainer_scalar_f` is for mixed scalar and `DataContainer` `a` and `b` in float.

epapoutsellis commented on Wrappers/Python/test/test_DataContainer.py:
```diff
@@ -931,6 +931,170 @@ def test_axpby7(self):
+        numpy.testing.assert_array_equal(res, d2.as_array())
+
+
+    def test_sapyb_datacontainer_c(self):
```
Rename to all_DataContainer_complex

epapoutsellis reviewed (changes requested):
Some renaming issues on the test methods.

paskino commented:
I updated the `sapyb` method with a new clearer variable `use_c_lib` instead of `do_numpy`. I added the example in the docstring.

I think the current names of the unittests are clearer than the ones proposed by @epapoutsellis

gfardell commented on Wrappers/Python/cil/framework/framework.py:
```diff
@@ -2481,7 +2481,61 @@ def maximum(self, x2, *args, **kwargs):
+        if out is None:
+            out = self * 0.
+            ret_out = True
+
```
The c implementation of `axpby` can work with `x` and `y` being arrays of floats or doubles, and `a` and `b` being either scalars or arrays  of floats or doubles respectively.

The dtype used is controlled by the dtype of `out`, and a `TypeError` is raised if this isn't float32 or float64
https://github.com/TomographicImaging/CIL/blob/0e78d20fa31f7ec90e18107ba5e4ef5a575d00d0/Wrappers/Python/cil/framework/framework.py#L2611

So I'd suggest creating `out`, then if dtype is float32 or float64 call `_axpyb` for the c code and otherwise reverting to python. This means we only pass to C the datatypes we are guaranteed to accept, whereas as it is it checks for complex only. 

We could also try to catch the runtime error of the C code and default to python if this happens?
https://github.com/TomographicImaging/CIL/blob/0e78d20fa31f7ec90e18107ba5e4ef5a575d00d0/Wrappers/Python/cil/framework/framework.py#L2637

The RunTime error  could be caused by anything :) but most likely memory access as we don't check on the array length passed - in which case numpy will fail too I assume.

I know you won't like changing requested behaviour with only a warning and not a crash though!

paskino replied:
It could still crash in the fallback method, no?

paskino commented:
Jenkins fails for numpy 1.15 and python 3.6

```
======================================================================
ERROR: test_sapyb_scalar_f_c (test_DataContainer.TestDataContainer)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/conda-bld/cil_1641920992965/test_tmp/Wrappers/Python/test/test_DataContainer.py", line 1119, in test_sapyb_scalar_f_c
    with self.assertRaises(numpy.core._exceptions.UFuncTypeError) as context:
AttributeError: module 'numpy.core' has no attribute '_exceptions'

======================================================================
ERROR: test_SPDHG_vs_SPDHG_explicit_axpby (test_algorithms.TestSPDHG)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/conda-bld/cil_1641920992965/test_tmp/Wrappers/Python/test/test_algorithms.py", line 1028, in test_SPDHG_vs_SPDHG_explicit_axpby
    algos[0].run(1000, verbose=0)
  File "/home/jenkins/conda-bld/cil_1641920992965/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python3.6/site-packages/cil/optimisation/algorithms/Algorithm.py", line 269, in run
    self.__next__()
  File "/home/jenkins/conda-bld/cil_1641920992965/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python3.6/site-packages/cil/optimisation/algorithms/Algorithm.py", line 117, in __next__
    self.update()
  File "/home/jenkins/conda-bld/cil_1641920992965/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python3.6/site-packages/cil/optimisation/algorithms/SPDHG.py", line 173, in update
    y_k.axpby(self.sigma[i], 1., self.y_old[i], out=y_k)
  File "/home/jenkins/conda-bld/cil_1641920992965/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python3.6/site-packages/cil/framework/framework.py", line 2545, in axpby
    self._axpby(a,b,y,out, dtype, num_threads)
  File "/home/jenkins/conda-bld/cil_1641920992965/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python3.6/site-packages/cil/framework/framework.py", line 2598, in _axpby
    ndy = ndy.astype(dtype, casting='safe')
TypeError: Cannot cast array from dtype('float64') to dtype('float32') according to the rule 'safe'

----------------------------------------------------------------------
Ran 475 tests in 347.033s
```

paskino commented:
I removed the SPDHG changes and made sure that the data passed to it in the unit test is float32 and now everything is happy. 

I believe it can be merged @epapoutsellis @gfardell ?
